### PR TITLE
ENG-4656: Dynamic role management

### PR DESCRIPTION
**Pull request by eugeniosant**

### CODE REVIEW

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java [hunk 1]
```diff
@@ -64,13 +65,15 @@ protected void insertAndCleanCache(Cache cache, Map<String, O> objects) {
     protected void insertAndCleanCache(Cache cache, Map<String, O> objects, String codesCacheKey, String cacheKeyPrefix) {
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java]: why we cannot put a check to not null and return an error? don't you think we are complicating too much the base common class (is it safe ?) to handle a case of a caller that wants just to clear the cache for a key? ( ... and should use another method ... )

File: engine/src/main/java/com/agiletec/aps/system/common/entity/ApsEntityManager.java [hunk 2]
```diff
@@ -91,28 +96,32 @@ public abstract class ApsEntityManager extends AbstractService
...
 
     private IEntityManagerCacheWrapper cacheWrapper;
-
+    
     @Override
     public void init() throws Exception {
         this.entityDom.setRootElementName(this.getXmlAttributeRootElementName());
         this.getCacheWrapper().initCache(super.getName());
-        logger.debug("{} : inizializated", this.getName());
+        this.refreshAttributeRoles();
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/entity/ApsEntityManager.java]: a refresh operation inside the `init`? ... and we have `refresh` method ... something strange ...

File: engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java [hunk 3]
```diff
@@ -61,7 +70,32 @@ private void loadDefaultRoles(String attributeRolesFileName, Map<String, Attribu
...
 		}
 	}
-	
+    
+    public void loadLocalRoles(String configItemName, Map<String, AttributeRole> defaultCollection) {
+        try {
+            ConfigInterface configManager = this.getBeanFactory().getBean(ConfigInterface.class);
+			String xml = configManager.getConfigItem(configItemName);
+            if (StringUtils.isBlank(xml)) {
+                return;
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java:81]: can a log at debug level with the value of `configItemName` be useful?

File: engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java [hunk 4]
```diff
@@ -61,7 +70,32 @@ private void loadDefaultRoles(String attributeRolesFileName, Map<String, Attribu
...
+            AttributeRoleDOM dom = new AttributeRoleDOM();
+			Map<String, AttributeRole> attributeRoles = dom.extractRoles(xml, "configItem:" + configItemName);
+			List<AttributeRole> roles = new ArrayList<>(attributeRoles.values());
+			for (int i = 0; i < roles.size(); i++) {
+				AttributeRole role = roles.get(i);
+				if (defaultCollection.containsKey(role.getName())) {
+					_logger.warn("You can't override existing attribute role : {} - {}", role.getName(), role.getDescription());
+				} else {
+					defaultCollection.put(role.getName(), role);
+					_logger.info("Added new attribute role : {} - {}",role.getName(), role.getDescription());
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java]: maybe this should be a log at debug level?

File: engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java [hunk 5]
```diff
@@ -61,7 +70,32 @@ private void loadDefaultRoles(String attributeRolesFileName, Map<String, Attribu
...
+				}
+                role.setLocal(true);
+			}
+		} catch (EntException | BeansException e) {
+			_logger.error("Error loading local attribute Roles", e);
```
Comment (entgigi) [inline engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java]: can be useful to log also some parameters? `configItemName`?

File: engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java (hunk 3 above)
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java:81]: Done

File: engine/src/main/java/com/agiletec/aps/system/common/entity/ApsEntityManager.java (hunk 2 above)
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/common/entity/ApsEntityManager.java]: done

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java (hunk 1 above)
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/common/AbstractGenericCacheWrapper.java]: Restored previous method

File: engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java (hunk 4 above)
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java]: Done

File: engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java (hunk 5 above)
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/common/entity/loader/AttributeRolesLoader.java]: Done

**Review by entgigi — APPROVED** [review]
LGTM

**Pull request merged**